interceptors/logging/interceptors.go

```diff
@@ -22,65 +22,132 @@ func extractFields(tags tags.Tags) Fields {
+	}
+	if !c.isServer {
+		logger = logger.With("component", "client")
+	}
+	// Attach the kind, protocol and type as a structured log entry.
+	logger = logger.With("kind", c.kind)
+	logger = logger.With("type", string(c.typ))
+	logger = logger.With("protocol", "grpc")
+
+	return logger
+}
+
+// PostCall logs the server/method name details and error if any.
 func (c *reporter) PostCall(err error, duration time.Duration) {
 	if !c.opts.shouldLog(interceptors.FullMethod(c.service, c.method), err) {
 		return
 	}
 	if err == io.EOF {
 		err = nil
 	}
+	logger := c.addMetaDetails(c.logger)
 	// Get optional, fresh tags.
-	logger := c.logger.With(extractFields(tags.Extract(c.ctx))...)
-
+	logger = logger.With(extractFields(tags.Extract(c.ctx))...)
 	code := c.opts.codeFunc(err)
 	logger = logger.With("grpc.code", code.String())
 	if err != nil {
-		logger = logger.With("error", fmt.Sprintf("%v", err))
+		logger = logger.With("grpc.error", fmt.Sprintf("%v", err))
+	}
+	logger = logger.With(c.opts.durationFieldFunc(duration)...)
+	logger.Log(c.opts.levelFunc(code), "finished call")
+}
+
+// PostMsgSend logs the details of the servingObject that is flowing out of the rpc.
+// resp object wrt server.
+// Log the details of the first request, skip if the object is response.
+func (c *reporter) PostMsgSend(resp interface{}, err error, duration time.Duration) {
+	if !c.opts.shouldLog(interceptors.FullMethod(c.service, c.method), err) {
+		return
+	}
+
+	// If the first message is logged skip the rest of the logging.
+	// If the serving object is response, skip the logging.
+	if c.firstResponseMessageLogged || !c.isServer {
+		return
+	}
+	c.firstResponseMessageLogged = true
+	logger := c.addMetaDetails(c.logger)
+	logger = logger.With(extractFields(tags.Extract(c.ctx))...)
+	logger = logger.With("grpc.recv.duration", duration.String())
+	code := c.opts.codeFunc(err)
+	if err != nil {
+		logger = logger.With("grpc.error", fmt.Sprintf("%v", err))
 	}
-	logger.With(c.opts.durationFieldFunc(duration)...).Log(c.opts.levelFunc(code), fmt.Sprintf("finished %s %s call", c.kind, c.typ))
+	logger.With(c.opts.durationFieldFunc(duration)...).Log(c.opts.levelFunc(code), "started call")
 }
 
-func (c *reporter) PostMsgSend(interface{}, error, time.Duration) {}
+// PostMsgReceive logs the details of the servingObject that is flowing into the rpc.
+// req object wrt server.
+// Log the details of the request, skip if the object is response.
+func (c *reporter) PostMsgReceive(req interface{}, err error, duration time.Duration) {
+	if !c.opts.shouldLog(interceptors.FullMethod(c.service, c.method), err) {
+		return
+	}
 
-func (c *reporter) PostMsgReceive(interface{}, error, time.Duration) {}
+	// If the first message for request is logged, skip the rest of the logging.
+	// If the serving object is a response, skip the logging.
+	if c.firstRequestMessageLogged || !c.isServer {
+		return
+	}
+	c.firstRequestMessageLogged = true
+	logger := c.addMetaDetails(c.logger)
```

Comment:
I think all those things can be abstracted in to one function (so  PostMsgReceive and Send can just inoke `logStarted`)

Reply:
Changed